sapirbitan/assignment4/assignment4.cpp

```diff
@@ -0,0 +1,112 @@
+#include <iostream>
+#include <vector>
+using namespace std;
+//this class contains the row and the column of the cell
+class Cell {
+public:
+	int row;
+	int column;
+};
+
+bool goodRowColumn(int row, int column, int num_of_rows, int num_of_columns) {
+	return (row >= 0 && row < num_of_rows && column >= 0 && column < num_of_columns);
+}
+//finds all of the adjacent cells
+vector<Cell> getNeighbors(int current_row, int current_column, int num_of_rows, int num_of_columns) {
+	vector<Cell> neighbors;
+	Cell temp_cell;
+	//top cell
+	if (goodRowColumn(current_row -1, current_column, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row -1;
+		temp_cell.column = current_column;
+		neighbors.push_back(temp_cell);
+	}
+	//bottum cell
+	if (goodRowColumn(current_row +1, current_column, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row + 1;
+		temp_cell.column = current_column;
+		neighbors.push_back(temp_cell);
+	}
+	//left cell
+	if (goodRowColumn(current_row, current_column-1, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row;
+		temp_cell.column = current_column-1;
+		neighbors.push_back(temp_cell);
+	}
+	//right cell
+	if (goodRowColumn(current_row, current_column + 1, num_of_rows, num_of_columns)) {
+		temp_cell.row = current_row;
+		temp_cell.column = current_column + 1;
+		neighbors.push_back(temp_cell);
+	}
+	return neighbors;//returns the valid adjacent cells
+}
+//checks if the island is bigger than one cell on the map
+void checkArea(int row, int column, vector<vector<bool>> &map, vector<vector<bool>> &been_there) {
+	if (!map[row][column])
+		return;
+	been_there[row][column] = true;//the only places we want to mark as been_there are the ones who are land
+								  //in order to know it belongs to the same island
+	vector<Cell> neighbors = getNeighbors(row, column,map.size(),map[0].size());
+	for (int i = 0; i < neighbors.size();i++) {
+		if(!(been_there[neighbors[i].row][neighbors[i].column]))
+			checkArea(neighbors[i].row, neighbors[i].column, map, been_there);
+	}
+}
+//finds out how many islands are there in the map
+int findLand(int num_of_rows, int num_of_columns, vector<vector<bool>> map) {
```

Comment:
You should pass all non-trivial objects that you don't plan to modify (here non-trivial means everything that occupies more than a couple of bytes in memory) as a `const &`. For example, in this case your map should be `const vector<vector<bool>>& map`.

`const` forces the function not to be able to modify the object and `&` means that your map will be referenced instead of copied on each function call. This makes the code run a lot faster if you have a lot of large objects that you have to pass around :)